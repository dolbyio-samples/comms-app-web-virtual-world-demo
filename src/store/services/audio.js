import VoxeetSDK from "@voxeet/voxeet-web-sdk";

export const startRemoteAudio = (participant) => {
  return VoxeetSDK.audio.remote
    .start(participant)
    .then((response) => ({ response }))
    .catch((error) => ({ error }));
};

export const stopRemoteAudio = (participant) => {
  return VoxeetSDK.audio.remote
    .stop(participant)
    .then((response) => ({ response }))
    .catch((error) => ({ error }));
};

export const startLocalAudio = () => {
  return VoxeetSDK.audio.local
    .start()
    .then((response) => ({ response }))
    .catch((error) => ({ error }));
};

export const stopLocalAudio = () => {
  return VoxeetSDK.audio.local
    .stop()
    .then((response) => ({ response }))
    .catch((error) => ({ error }));
};
