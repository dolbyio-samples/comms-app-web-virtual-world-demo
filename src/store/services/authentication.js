export const retrieveAccessToken = () => async () => {
  const authUrl = import.meta.env.VITE_AUTH_URL;

  const response = await fetch(authUrl);
  const { access_token } = await response.json();
  return access_token;
};
