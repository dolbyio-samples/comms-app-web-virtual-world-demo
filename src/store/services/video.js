import VoxeetSDK from "@voxeet/voxeet-web-sdk";

export const startVideo = (participant) => {
  return VoxeetSDK.conference
    .startVideo(participant, {
      audio: true,
      video: true,
    })
    .then((response) => ({ response }))
    .catch((error) => ({ error }));
};

export const stopVideo = (participant) => {
  return VoxeetSDK.conference
    .stopVideo(participant)
    .then((response) => ({ response }))
    .catch((error) => ({ error }));
};
