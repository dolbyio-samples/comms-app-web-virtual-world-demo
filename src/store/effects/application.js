import { createEffect } from "../utils/store";

export const sdkInitialized = createEffect("SDK_INITIALIZED");

export const setLocalParticipantID = createEffect("SET_LOCAL_PARTICIPANT_ID");
export const setCurrentConferenceID = createEffect("SET_CURRENT_CONFERENCE_ID");

export const setCurrentDeviceID = createEffect("SET_CURRENT_DEVICE_ID");

export const leaveConference = createEffect("LEAVE_CONFERENCE");

export const addMessageID = createEffect("ADD_MESSAGE_ID");
export const removeMessageID = createEffect("REMOVE_MESSAGE_ID");
