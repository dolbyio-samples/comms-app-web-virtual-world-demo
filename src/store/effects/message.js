import { createEffect } from "../utils/store";

export const addMessage = createEffect("ADD_MESSAGE");
export const removeMessage = createEffect("REMOVE_MESSAGE");
