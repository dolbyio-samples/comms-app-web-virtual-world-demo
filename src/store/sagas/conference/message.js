import { createId } from "@paralleldrive/cuid2";
import { addMessage } from "../../effects/message";
import { application } from "../../effects";
import { put, call } from "redux-saga/effects";

export function* displayMessage({ type, message }) {
  const id = yield call(createId);
  yield put(
    addMessage({
      entityType: "messages",
      data: {
        [id]: {
          type,
          message,
        },
      },
    })
  );

  yield put(application.addMessageID({ id }));
}
