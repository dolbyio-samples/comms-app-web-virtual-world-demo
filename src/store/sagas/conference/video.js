import { call, select } from "redux-saga/effects";

import { getEntity } from "../../reducers/entities/selectors";
import { startVideo as start, stopVideo as stop } from "../../services/video";
import { displayMessage } from "./message";

export const getParticipant = (state, id) =>
  getEntity(state, "participants", id);

export function* startVideo({ payload }) {
  const participant = yield select(getParticipant, payload);
  if (!participant) return;

  const { error } = yield call(start, participant);
  if (error) {
    yield call(displayMessage, {
      type: "error",
      message: "could not start video",
    });
  }
}

export function* stopVideo({ payload }) {
  const participant = yield select(getParticipant, payload);
  if (!participant) return;

  const { error } = yield call(stop, participant);
  if (error) {
    yield call(displayMessage, {
      type: "error",
      message: "could not stop video",
    });
  }
}
