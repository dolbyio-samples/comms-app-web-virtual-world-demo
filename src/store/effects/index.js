import * as application from "./application";
import * as conference from "./conference";
import * as session from "./session";
import * as authentication from "./authentication";
import * as spatial from "./spatial";
import * as message from "./message";

export { authentication, application, conference, session, spatial, message };
