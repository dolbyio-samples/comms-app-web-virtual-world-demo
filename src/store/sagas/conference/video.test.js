import { stopVideo, startVideo, getParticipant } from "./video";
import { call, select } from "redux-saga/effects";
import { startVideo as start, stopVideo as stop } from "../../services/video";
import { vi, describe, it, expect } from "vitest";
import { displayMessage } from "./message.js";

vi.mock("@voxeet/voxeet-web-sdk", () => ({
  default: {},
}));

describe("video saga test suite", () => {
  it("should return a participant", () => {
    const state = {
      entities: {
        participants: {
          1: {
            name: "toto",
          },
        },
      },
    };

    const id = "1";

    const participant = getParticipant(state, id);

    expect(participant).toMatchSnapshot();
  });

  it("should not call start audio service", () => {
    const action = { type: "START_VIDEO", payload: "123" };
    const generator = startVideo(action);

    expect(generator.next().value).toEqual(
      select(getParticipant, action.payload)
    );

    expect(generator.next().done).toBeTruthy();
  });

  it("should call start audio service", () => {
    const action = { type: "START_VIDEO", payload: "123" };
    const generator = startVideo(action);

    expect(generator.next().value).toEqual(
      select(getParticipant, action.payload)
    );

    const participant = {};

    expect(generator.next(participant).value).toEqual(call(start, participant));
    expect(generator.next({ error: undefined }).done).toBeTruthy();
  });

  it("should call start audio service but fail and dispatch an action", () => {
    const action = { type: "START_VIDEO", payload: "123" };
    const generator = startVideo(action);

    expect(generator.next().value).toEqual(
      select(getParticipant, action.payload)
    );

    const participant = {};

    expect(generator.next(participant).value).toEqual(call(start, participant));
    expect(generator.next({ error: {} }).value).toEqual(
      call(displayMessage, {
        type: "error",
        message: "could not start video",
      })
    );
    expect(generator.next().done).toBeTruthy();
  });

  it("should not call stop video service", () => {
    const action = { type: "STOP_VIDEO", payload: "123" };
    const generator = stopVideo(action);

    expect(generator.next().value).toEqual(
      select(getParticipant, action.payload)
    );

    expect(generator.next().done).toBeTruthy();
  });

  it("should call stop video service", () => {
    const action = { type: "START_AUDIO", payload: "123" };
    const generator = stopVideo(action);

    expect(generator.next().value).toEqual(
      select(getParticipant, action.payload)
    );

    const participant = {};

    expect(generator.next(participant).value).toEqual(call(stop, participant));
    expect(generator.next({ error: undefined }).done).toBeTruthy();
  });

  it("should call stop video service but fail and dispatch an action to store", () => {
    const action = { type: "START_AUDIO", payload: "123" };
    const generator = stopVideo(action);

    expect(generator.next().value).toEqual(
      select(getParticipant, action.payload)
    );

    const participant = {};

    expect(generator.next(participant).value).toEqual(call(stop, participant));
    expect(generator.next({ error: {} }).value).toEqual(
      call(displayMessage, {
        type: "error",
        message: "could not stop video",
      })
    );
    expect(generator.next().done).toBeTruthy();
  });
});
