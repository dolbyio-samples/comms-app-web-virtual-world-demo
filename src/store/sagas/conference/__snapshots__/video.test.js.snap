// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`video saga test suite > should return a participant 1`] = `
{
  "name": "toto",
}
`;
